Support AWS CodeCommit as pipeline source repository

ybezsonov opened the pull request:
*Issue #, if available: #402

*Description of changes:*
Added support of AWS CodeCommit as repository source for pipeline-stack

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

shapirov103 commented on lib/pipelines/code-pipeline.ts:
```diff
@@ -48,10 +69,15 @@ export type PipelineProps = {
 
     /**
-     * Repository for the pipeline
+     * Repository for the pipeline (GitHub handle).
```
I assume you wanted to call out that this repository is a gitHub repo. Let's rephrase to state `GitHub repository for the pipeline`.

shapirov103 commented on lib/pipelines/code-pipeline.ts:
```diff
@@ -48,10 +69,15 @@ export type PipelineProps = {
+    /**
+     * CodeCommit repository properties for the pipeline (CodeCommit handle).
+     */
+    codeCommitProps: CodeCommitProps;
```
Looks like both codeCommitProps and repository fields are required while they are mutually exclusive.

shapirov103 commented on lib/pipelines/code-pipeline.ts:
```diff
@@ -32,6 +33,26 @@ export interface GitHubSourceRepository extends Omit<ApplicationRepository, "cre
+/**
+ * Props for the CodeCommit repository.
+ */
+export type CodeCommitProps = {
```
For consistency with the GitHub repo let's create an interface `CodeCommitSourceRepository` with properties `codeCommitRepoName`. 

This can enable us to declare the `repository` field in the PipelineProps as:

```
repository: GitHubSourceRepository | CodeCommitSourceRepository
```

This field will be required and remove the need to check when both are specified.

shapirov103 reviewed (commented):
@ybezsonov Great PR, thank you for your contribution!

A few comments for you. 

A couple of things to add to make it complete:

1. Let's add a unit test similar to [this](https://github.com/aws-quickstart/cdk-eks-blueprints/blob/main/test/stacks.test.ts#L96) otherwise we have no coverage. Ideally we could also add a pattern, but we can defer it until this version is published.
2. Let's add a small paragraph on codecommit support [here](https://github.com/aws-quickstart/cdk-eks-blueprints/blob/main/docs/pipelines.md#creating-a-pipeline) to let the customers know that we have this support.

ybezsonov commented:
Hi @shapirov103 
I added owner property to GitHubSourceRepository because I think it belongs to this interface.
I left PipelineProps.owner for backward compatibility, but made it optional. It can be deprecated and removed in next releases.
If both PipelineProps.owner and PipelineProps.repository.owner are defined, then PipelineProps.repository.owner will be used.

I added documentation and tests for several cases.

shapirov103 commented:
@ybezsonov when you get a chance, could you please look into lint failure? see the log above
https://github.com/aws-quickstart/cdk-eks-blueprints/runs/6997178622?check_suite_focus=true

ybezsonov commented:
> @ybezsonov when you get a chance, could you please look into lint failure? see the log above https://github.com/aws-quickstart/cdk-eks-blueprints/runs/6997178622?check_suite_focus=true

@shapirov103 I updated the code, hope it helps.

shapirov103 commented on lib/pipelines/code-pipeline.ts:
```diff
@@ -255,19 +290,35 @@ export class ApplicationStage extends cdk.Stage {
+        } else {
+            let gitHubRepo = props.repository as GitHubSourceRepository;
+            let githubProps : cdkpipelines.GitHubSourceOptions | undefined = undefined;
+            let gitHubOwner = gitHubRepo.owner ? gitHubRepo.owner : gitHubRepo.owner;
```
@ybezsonov this looks like a defect: I think you meant to use `gitHubRepo.owner` if defined and fall back to  `props.owner` otherwise. Otherwise, it will be a breaking change.

Please consider nullish coalscing for string defaulting:

```typescript
const gitHubOwner = gitHubRepo.owner ?? props.owner;
```

ybezsonov replied:
It was copy/paste error, but quite dangerous one.
Thank you for spotting this!

shapirov103 commented on lib/pipelines/code-pipeline.ts:
```diff
@@ -255,19 +290,35 @@ export class ApplicationStage extends cdk.Stage {
             pipelineName: props.name,
             synth: new cdkpipelines.ShellStep(`${props.name}-synth`, {
-              input: cdkpipelines.CodePipelineSource.gitHub(`${props.owner}/${props.repository.repoUrl}`, branch, githubProps), 
+              input: codePipelineSource,
```
:+1:

shapirov103 reviewed (commented):
@ybezsonov almost there. Was about to merge but spotted a defect. Tiny change but potentially big impact.

shapirov103 reviewed (approved):
LGTM
(will be included in the next minor revision update - 1.1.0)
